services/navigation/builtin/builtin.go

```diff
@@ -419,16 +459,58 @@ func (svc *builtIn) startWaypointExperimental(extra map[string]interface{}) erro
 		}
 
 		// loop until no waypoints remaining
-		for wp, err := svc.nextWaypoint(svc.cancelCtx); err == nil; wp, err = svc.nextWaypoint(svc.cancelCtx) {
+		for wp, err := svc.nextWaypoint(cancelCtx); err == nil; wp, err = svc.nextWaypoint(cancelCtx) {
+			if cancelCtx.Err() != nil {
+				return
+			}
+
+			navOnceCancelCtx := svc.resetForNavOnce(cancelCtx, wp)
 			svc.logger.Infof("navigating to waypoint: %+v", wp)
-			if err := navOnce(svc.cancelCtx, wp); err != nil {
+			if err := navOnce(navOnceCancelCtx, wp); err != nil {
+				if svc.waypointIsDeleted() {
+					svc.logger.Infof("skipping waypoint %+v since it was deleted", wp)
+					continue
+				}
+
 				svc.logger.Infof("skipping waypoint %+v due to error while navigating towards it: %s", wp, err)
+				if err := svc.waypointReached(cancelCtx); err != nil {
+					svc.logger.Info("can't mark waypoint %+v as reached, exiting navigation due to error: %s", wp, err)
+				}
 			}
 		}
 	})
-	return nil
+}
+
+func (svc *builtIn) resetForNavOnce(ctx context.Context, wp navigation.Waypoint) context.Context {
+	svc.mu.Lock()
+	defer svc.mu.Unlock()
+	svc.waypointInProgress = &wp
+	navOnceCancelCtx, fn := context.WithCancel(ctx)
+	svc.navOnceCancelFunc = fn
+	return navOnceCancelCtx
+}
+
+func (svc *builtIn) waypointIsDeleted() bool {
+	svc.mu.RLock()
+	defer svc.mu.RUnlock()
+	return svc.waypointInProgress == nil
 }
 
 func (svc *builtIn) GetObstacles(ctx context.Context, extra map[string]interface{}) ([]*spatialmath.GeoObstacle, error) {
+	svc.mu.RLock()
+	defer svc.mu.RUnlock()
 	return svc.obstacles, nil
 }
+
+func (svc *builtIn) deleteAllWaypoints(ctx context.Context) error {
```

Comment:
Is this used anywhere except in the tests?

If not, it shouldn't be a function on the service struct; just make it a helper function in the test file itself.